iceoryx2-cal/src/resizable_shared_memory/mod.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! A [`ResizableSharedMemory`] is identified by a name and allows multiple processes to share
+//! memory between them (inter-process memory). One process owns the [`ResizableSharedMemory`]
+//! which can be created via the [`ResizableSharedMemoryBuilder`] and many processes can have a
+//! [`ResizableSharedMemoryView`] that can be constructed via [`ResizableSharedMemoryViewBuilder`].
+//!
+//! The [`ResizableSharedMemoryView`] never owns the [`ResizableSharedMemory`] and has only
+//! read-only access to it while the [`ResizableSharedMemory`] can use
+//! [`ResizableSharedMemory::allocate()`] to acquire memory and distribute it between the
+//! [`ResizableSharedMemoryView`]s.
+//!
+//! Whenever the [`ResizableSharedMemoryView`] receives an offset it must be registered via
+//! [`ResizableSharedMemoryView::register_and_translate_offset()`] and unregistered via
+//! [`ResizableSharedMemoryView::unregister_offset()`]. As soon as the [`ResizableSharedMemory`]
+//! calls [`ResizableSharedMemory::deallocate()`] unused [`SharedMemory`] segments may be recycled.
+//!
+//! # Example
+//!
+//! ```
+//! // owner of the ResizableSharedMemory
+//! use iceoryx2_cal::resizable_shared_memory::*;
+//! use iceoryx2_bb_system_types::file_name::FileName;
+//! use iceoryx2_cal::shm_allocator::ShmAllocator;
+//! use iceoryx2_cal::shared_memory::SharedMemory;
+//! use iceoryx2_cal::named_concept::*;
+//! use std::alloc::Layout;
+//! use std::time::Duration;
+//!
+//! fn example<Alloc: ShmAllocator, Shm: SharedMemory<Alloc>, Memory: ResizableSharedMemory<Alloc, Shm>>(
+//!     name: &FileName
+//! ) {
+//!     // owner process creates a new memory
+//!     let memory = Memory::MemoryBuilder::new(name)
+//!         // hint to the underlying allocator that we need up to 128 chunks of memory
+//!         //
+//!         // note: as soon as there are more than 128 chunks requested a new resized segment is
+//!         //       created
+//!         .max_number_of_chunks_hint(128)
+//!         // hint to the underlying allocator that the chunks are not exceeding the Layout of
+//!         // [`u16`]
+//!         //
+//!         // note: as soon as there is a chunk requested that exceeds the provided layout hint a
+//!         //       new resized segment is created
+//!         .max_chunk_layout_hint(Layout::new::<u32>())
```

Comment:
Do you mean `[u32]`?

Reply:
Nope, this is one layer lower and here you can mix types as you want. First you can give a hint for a `u32` than for `[Whatever]` if you like.